server-common/src/main/java/org/apache/kafka/server/fault/ProcessTerminatingFaultHandler.java

```diff
@@ -0,0 +1,97 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License. You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.kafka.server.fault;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import org.apache.kafka.common.utils.Exit;
+
+
+/**
+ * This is a fault handler which terminates the JVM process.
+ */
+final public class ProcessTerminatingFaultHandler implements FaultHandler {
+    private static final Logger log = LoggerFactory.getLogger(ProcessTerminatingFaultHandler.class);
+
+    private final Runnable action;
+    private final boolean shouldHalt;
+
+    private ProcessTerminatingFaultHandler(boolean shouldHalt, Runnable action) {
+        this.action = action;
+        this.shouldHalt = shouldHalt;
+    }
+
+    @Override
+    public RuntimeException handleFault(String failureMessage, Throwable cause) {
+        if (cause == null) {
+            log.error("Encountered fatal fault: {}", failureMessage);
+        } else {
+            log.error("Encountered fatal fault: {}", failureMessage, cause);
+        }
+
+        try {
+            action.run();
+        } catch (Throwable e) {
+            log.error("Failed to run terminating action.", e);
+        }
+
+        int statusCode = 1;
+        if (shouldHalt) {
+            Exit.halt(statusCode);
+        } else {
+            Exit.exit(statusCode);
+        }
+
+        return null;
+    }
+
+    /**
+     * Same as {@code exitingWithAction} with a no-op action.
+     */
+    public static FaultHandler exiting() {
+        return exitingWithAction(() -> { });
+    }
+
+    /**
+     * Handle faults by running an action and calling {@code Exit.exit}.
+     *
+     * The default implementation of {@code Exit.exit} calls {@code Runtime.exit} which
+     * waits on all of the shutdown hooks executing.
+     */
+    public static FaultHandler exitingWithAction(Runnable action) {
```

Comment:
This feels awkward to me. If it's creating a new object "create", "build", etc. should be somewhere in the name. Like `createExitingFaultHandlerWithAction`. But that's a long name. Plus now we have a combinatorial explosion of builder functions. createWithXAndY, createWithXAndYAndZ. Not good.

If you want to make this nicer than just exposing the constructor then just make a builder. `ProcessTerminatingFaultHandler.Builder().setHalt(true).build()` is easy to read and makes it easy to add new arguments if we want to.

Reply:
Sounds good. I added a builder pattern.